src/classes/CW_CampaignTriggerHandlerTest.cls

```diff
@@ -0,0 +1,266 @@
+		List<CW_Schedule__c> updatedSchedules = [select Id, Campaign_Stage__c from CW_Schedule__c where Id in:scheduleIds];
+		
+		for (CW_Schedule__c s :updatedSchedules) system.assertEquals('stage 2', s.Campaign_Stage__c);
+	}
+	
+	static testMethod void testRecurrenceType()
+	{
+		CW_Campaign__c camp = TestDataUtil.createCampaign();	// insert with no recurrence to avoid triggers
+    	insert camp;
+    	camp.Recurrence_Type__c = ConstantUtil.CAMPAIGN_RECURRENCE_TYPE_RECURRING;
+    	
+    	CW_Campaign__c campNew = camp.clone(true,true);
+    	campNew.Recurrence_Type__c = ConstantUtil.CAMPAIGN_RECURRENCE_TYPE_SINGLE;
+    	
+    	CW_Schedule__c sch = TestDataUtil.createSchedule(camp);
+    	insert sch;
+    	
+    	Test.startTest();
+    	
+    	CW_CampaignTriggerHandler cth = new CW_CampaignTriggerHandler(true);
+    	cth.cascadeDeleteSchedulesOnUpdate(new List<CW_Campaign__c>{campNew}, new Map<Id,CW_Campaign__c>(new List<CW_Campaign__c>{camp}));
+    	
+    	Test.stopTest();
+    	
+    	// assert that new schdule was created and old one deleted
+    	system.assertEquals(1,[select count() from CW_Schedule__c where id != :sch.id]);
+    	system.assertEquals(0,[select count() from CW_Schedule__c where id = :sch.id]);
+	}
+	
+	static testMethod void testUpdateCampaignStageWhenBriefIsRejectedOrRecalled()
+	{
+		CW_Campaign__c camp = TestDataUtil.createCampaign();	// insert with no recurrence to avoid triggers
+    	camp.Recurrence_Type__c = ConstantUtil.CAMPAIGN_RECURRENCE_TYPE_RECURRING;
+    	camp.Brief_Approval_Status__c = 'In Approval';
+    	insert camp;
+    	
+    	Test.startTest();
+    	camp.Brief_Approval_Status__c = 'Rejected';
+    	update camp;
+    	Test.stopTest();
+    	system.assertEquals('Draft',[Select Stage__c from CW_Campaign__c where Id=:camp.Id].Stage__c);
+    	
+    	camp.Brief_Approval_Status__c = 'In Approval';
+    	camp.Cost_Approval_Status__c = 'Approved';
+    	camp.Roadmap_Approval_Status__c = 'Approved';
+    	update camp;
+    	
+    	camp.Brief_Approval_Status__c = 'Rejected';
+    	update camp;
+    	system.assertEquals('Roadmap Approved',[Select Stage__c from CW_Campaign__c where Id=:camp.Id].Stage__c);
+    	
+    	//recall brief for an urgent campaign -> camp stage should revert to draft
+    	camp.Brief_Approval_Status__c = 'In Approval';
+    	camp.Roadmap_Approval_Status__c = 'Not Submitted';
+    	update camp;
+    	
+    	camp.Brief_Approval_Status__c = 'Not Submitted';
+    	update camp;
+    	system.assertEquals('Draft',[Select Stage__c from CW_Campaign__c where Id=:camp.Id].Stage__c);
+    	//recall the brief when roadmap is approved -> camp stage should revert to roadmap approved
+    	camp.Brief_Approval_Status__c = 'In Approval';
+    	camp.Roadmap_Approval_Status__c = 'Approved';
+    	update camp;
+    	
+    	camp.Brief_Approval_Status__c = 'Not Submitted';
+    	update camp;
+    	system.assertEquals('Roadmap Approved',[Select Stage__c from CW_Campaign__c where Id=:camp.Id].Stage__c);
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.